### DISCUSSION TITLE
What is the meaning of '[..]'?

### DISCUSSION THREAD

**Discussion by bentxt** [toitlang/toit#14] [Q&A] [answered]
Hi all

I just read this declaration in a library file

`operator [..] --from=0 --to=size -> string:
`
What is the meaning of '[..]'? 

Its a bit hard to search for.

Thanks
Ben

**Comment by floitsch** [comment, marked as answer]
It the "slice" operator of Toit.

When the compiler sees `o[x..y]` (where both `x` and `y` can be omitted), then that expression resolves to the operator `[..] --from=x --to=y`.

It is used in strings and lists to get subsections of them:

```
main:
  str := "1234"
  sub := str[1..]  // => "234"
  sub = str[2..3]  // => "3"

  list := [1, 2, 3, 4]
  sub_list := list[..2]  // => [1, 2, 3]
```

Note that slices are generally implemented as views into the object. This means that a slice of a big string might keep the string alive.
Similarly, a slice of a list is a view into the list. Changing one, affects the other.
